_posts/2025-04-28-sparse-autodiff.md

```diff
@@ -664,5 +664,222 @@ This symmetry enables **colorings with fewer colors**, reducing the complexity o
+
+<aside class="l-body box-note" markdown="1">
+The authors of this blog post are all developers of the ASD ecosystem in Julia. We are not aware of a similar ecosystem in Python or R, which is why we chose Julia to present it.
+</aside>
+
+### Necessary packages
+
+Here are the packages we will use for this demonstration.
+We also use a few other packages for visualization.
+
+| Package | Purpose |
+|---|---|
+| [SparseConnectivityTracer.jl](https://github.com/adrhill/SparseConnectivityTracer.jl) | Sparsity pattern detection with operator overloading. |
+| [SparseMatrixColorings.jl](https://github.com/gdalle/SparseMatrixColorings.jl) | Greedy algorithms for colorings, decompression utilities. |
+| [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) | Forward-mode AD and computation of JVPs. |
+| [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl) | High-level interface bringing all of these together. |
+
+As with any language, the first step is importing the dependencies.
+
+```julia
+using DifferentiationInterface  # no prefix necessary
+import SparseConnectivityTracer  # requires prefix
+import SparseMatrixColorings
+import ForwardDiff
+```
+
+### Test function
+
+As our test function, we choose a very simple iterated difference operator.
+It takes a vector $\mathbf{x} \in \mathbb{R}^n$ and outputs a slightly shorter vector $y \in \mathbb{R}^{n-k}$ depending on the number of iterations $k$.
+In pure Julia, this is written as follows:
+
+```julia
+
+function differences(x)
+    n = length(x)
+    y = similar(x, n - 1)
+    for i in 1:(n - 1)
+        y[i] = x[i + 1] - x[i]
+    end
+    return y
+end
+
+function iterated_differences(x, k)
+    if k == 0
+        return x
+    else
+        y = iterated_differences(x, k - 1)
+        return differences(y)
+    end
+end
+```
+
+Let us check that the function returns what we expect:
+
+```julia
+julia> iterated_differences([1, 4, 9, 16], 1)
+3-element Vector{Int64}:
+ 3
+ 5
+ 7
+
+julia> iterated_differences([1, 4, 9, 16], 2)
+2-element Vector{Int64}:
+ 2
+ 2
+```
+
+### Dense Jacobian
+
+The key concept behind DifferentiationInterface.jl is that of "backend".
+There are several AD systems in Julia, with different features and tradeoff, but we access them through a common API.
+Here, we use ForwardDiff.jl as our dense backend:
+
+```julia
+dense_backend = DI.AutoForwardDiff()
+```
+
+To build a sparse backend, we bring together three ingredients corresponding to the various phases of ASD:
+
+```julia
+sparse_backend = DI.AutoSparse(
+    dense_backend;  # to compute JVPs
+    sparsity_detector=SCT.TracerSparsityDetector(),
+    coloring_algorithm=SMC.GreedyColoringAlgorithm(),
+)
+```
+
+We can now obtain the Jacobian of `iterated_differences` (with respect to $\mathbf{x}$) using either backend, and compare the results:
+
+```julia
+julia> x, k = rand(10), 3;
+
+julia> DI.jacobian(iterated_differences, dense_backend, x, DI.Constant(k))
```

Comment:
Would be nice if our example didn't have to use `DI.Constant(k)`, as it adds some complexity.

Reply:
It's either that or we make a callable struct where the number of iterations k is encoded. I thought this would  be simpler, and also showcase the features of DI